opentelemetry-exporter-cloud-trace/src/opentelemetry/tools/resource_detector.py

```diff
@@ -42,5 +76,7 @@ def detect(self) -> "Resource":
             self.cached = True
             for resource_finder in _RESOURCE_FINDERS:
                 found_resources = resource_finder()
-                self.gcp_resources.update(found_resources)
+                if found_resources:
+                    self.gcp_resources = found_resources
+                    break
```

Comment:
If I understand this correctly, get_gke_resources() will run first. If it returns something, then we're done. If it returns nothing, then we'll try get_gce_resources() instead. And this only works because of the ordering of _RESOURCE_FINDERS.

This seems likely to break in future code changes, because it doesn't seem intuitive that the order of _RESOURCE_FINDERS should matter. Can you add a comment to _RESOURCE_FINDERS to explain why the order is important?

Reply:
Alternatively, I see that the extraction of fields like host.id and cloud.zone are duplicated in the two resource finders. Perhaps there's a way to restructure the code? Something like:

```
def get_google_resources()
  all_metadata = _get_all_google_metadata()
  resources = {}
  # add common resources to the map here
  if not container_name or not pod_namespace:
    # add gke specific resources to the map here
  return resources
```

Reply:
I'd like to avoid making one large function as Cloud Monitoring might add support for more Google tools in the future, leading to one huge function. However, it does seem that `host.id` and `cloud.zone` are duplicated, I'll extract that out.